### PR TITLE
float exponentiation for gates

### LINKED ISSUE #2805: Gate.power()
A method on the Gate class to raise it to a power.

The base implementation is easy. Corner cases:

1 -> self
0 -> Identity (I'm not sure about this as identity currently implies wait).
-1 -> inverse
-2 -> inverse().power(2)
0.5 -> square root of gate

### LINKED ISSUE #2810: gate.sqrt()
### What is the expected enhancement?

Similar to #2805,  a `gate.root(<int>)` is needed. This will allow `gate.power` to support rational numbers as exponent.

### PR DESCRIPTION
**Pull request by 1ucian0**
Fixes #2805
Fixes #2810 

~This PR allows integer exponentiation using `instruction.gate`. I did not fully implement rational exponentiation because a `root` method is needed (#2810). Adding it here it seems too intrusive (and I'm not fully sure how to do it :-/)~

### CODE REVIEW

File: qiskit/circuit/instruction.py [hunk 1]
```diff
@@ -331,3 +332,29 @@ def broadcast_arguments(self, qargs, cargs):
             flat_qargs = [qarg for sublist in qargs for qarg in sublist]
             flat_cargs = [carg for sublist in cargs for carg in sublist]
             yield flat_qargs, flat_cargs
+
+    def power(self, exponent):
+        """
+        Creates an instruction with the gate repeated `exponent` amount of times.
+        Negative exponent creates an instruction with Gate.inverse().
+        Zero exponent creates an empty instruction.
```
[the remark this answers is not in the document]
Comment (1ucian0) [reply qiskit/circuit/instruction.py]: I agree that we need a gate to say 'nothing, not even wait'. The empty instruction could do this... I can check what's the unroller doing in that case.

File: qiskit/circuit/instruction.py [hunk 2]
```diff
@@ -331,3 +332,29 @@ def broadcast_arguments(self, qargs, cargs):
...
+
+        Args:
+            exponent (number): Gate^exponent
```
[the remark this answers is not in the document]
Comment (1ucian0) [reply qiskit/circuit/instruction.py]: Yes...
```suggestion
            exponent (int): Gate^exponent
```

File: qiskit/circuit/instruction.py [hunk 3]
```diff
@@ -331,3 +332,29 @@ def broadcast_arguments(self, qargs, cargs):
...
+
+        Returns:
+            Instruction: Containing the definition.
+        """
+        instruction = Instruction(name="%s^%s" % (self.name, exponent),
+                                  num_qubits=self.num_qubits,
+                                  num_clbits=self.num_clbits,
+                                  params=self.params)
+        qargs = [] if self.num_qubits == 0 else QuantumRegister(self.num_qubits, 'q')
```
[the remark this answers is not in the document]
Comment (1ucian0) [reply qiskit/circuit/instruction.py]: gates dont have `qargs` and `cargs` associated anymore...

File: qiskit/circuit/instruction.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (1ucian0) [reply qiskit/circuit/instruction.py]: indeed.. unroller does not know how to handle empty instruction. I will add this feature in a different PR.

File: qiskit/circuit/instruction.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (kdk) [reply qiskit/circuit/instruction.py]: Oh, in self.definition they do, but maybe we're supporting here even gates without a definition?

File: qiskit/circuit/instruction.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (1ucian0) [reply qiskit/circuit/instruction.py]: In a291c72b, `power(0)` means `u3(0,0,0)`.

File: qiskit/circuit/instruction.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (1ucian0) [reply qiskit/circuit/instruction.py]: `self.defintions` are like subcircuits, that's way they have `qargs` and `cargs`. A gate, by itself, does not. And `power` is on the gate.

File: qiskit/circuit/instruction.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (ajavadia) [reply qiskit/circuit/instruction.py]: so i guess update the docstring to say `power(0)` creates `u3(0,0,0)` instructions on all qubits, which are akin to `nop`?
I'm not sure we need an empty instruction.

[the remark this answers is not in the document]
Comment (1ucian0) [reply qiskit/circuit/instruction.py]: `u3(0,0,0)` is a good nop. that will make if for now.

[the remark this answers is not in the document]
Comment (1ucian0) [reply qiskit/circuit/instruction.py]: Docstring fixed in 3aa2bde5

File: test/python/circuit/test_instruction_power.py [hunk 4]
```diff
@@ -0,0 +1,225 @@
...
+    @data(2, 3, 4, 5, 6, 7, 8, 9)
+    def test_direct_root(self, degree):
+        """Test nth root"""
+        result = SGate().power(1 / degree)
+        self.assertEqual(result.name, 's^'+str(1/degree))
+        self.assertEqual(len(result.definition), 1)
+        assert_allclose(matrix_power(result.definition[0][0].to_matrix(), degree),
+                        SGate().to_matrix())
+
+    @data(2.1, 3.2, 4.3, 5.4, 6.5, 7.6, 8.7, 9.8)
```
Comment (kdk) [inline test/python/circuit/test_instruction_power.py]: Do we have a test for floats < 1, negative?

File: test/python/circuit/test_instruction_power.py [hunk 5]
```diff
@@ -0,0 +1,225 @@
...
+    def test_float_gt_one(self, exponent):
+        """Test greater-than-one exponents """
+        result = SGate().power(exponent)
+        self.assertEqual(result.name, 's^'+str(exponent))
+        self.assertEqual(len(result.definition), 1)
+        assert_allclose(SGate().to_matrix() ** exponent, result.definition[0][0].to_matrix())
```
Comment (kdk) [inline test/python/circuit/test_instruction_power.py]: Might be worth a comment that `**` works here only because `SGate` is diagonal.

File: qiskit/circuit/instruction.py [hunk 6]
```diff
@@ -307,7 +307,49 @@ def broadcast_arguments(self, qargs, cargs):
...
+        cargs = [] if self.num_clbits == 0 else ClassicalRegister(self.num_clbits, 'c')
+
+        if exponent > 0:
+            sub_instruction = (self, qargs[:], cargs[:])
+        elif exponent < 0:
+            sub_instruction = (self.inverse(), qargs[:], cargs[:])
+        else:
+            from qiskit.extensions.standard.u3 import U3Gate  # pylint: disable=cyclic-import
+            instruction.definition = [(U3Gate(0, 0, 0), qargs[:], cargs[:])]
```
[the remark this answers is not in the document]
Comment (1ucian0) [reply qiskit/circuit/instruction.py]: Indeed. Added in 54d482e7. Thanks!

File: qiskit/circuit/instruction.py [hunk 7]
```diff
@@ -307,7 +307,49 @@ def broadcast_arguments(self, qargs, cargs):
...
+            Instruction: Containing the definition.
+
+        Raises:
+            QiskitError: If called by a non-integer exponent.
+        """
+        if int(exponent) != exponent:
+            raise QiskitError("Instruction only supports integer exponentiation.")
+        exponent = int(exponent)
+
+        instruction = Instruction(name="%s^%s" % (self.name, exponent),
```
[the remark this answers is not in the document]
Comment (1ucian0) [reply qiskit/circuit/instruction.py]: `type(self)(...)` where `self` is a `CnotGate` would return `CnotGate`. The most generic type we have is `Instruction` and we use it for custom composite gate, right?

File: qiskit/circuit/gate.py [hunk 8]
```diff
@@ -42,6 +47,28 @@ def to_matrix(self):
...
+    def power(self, exponent):
+        if int(exponent) == exponent:
+            return super().power(exponent)
+
+        from qiskit.extensions.unitary import UnitaryGate  # pylint: disable=cyclic-import
+        # Should be diagonalized because it's a unitary.
+        decomposition, unitary = schur(self.to_matrix(), output='complex')
+        # Raise the diagonal entries to the specified power
+        decomposition_power = list()
+        for element in decomposition.diagonal():  # assert off-diagonal are 0
```
[the remark this answers is not in the document]
Comment (1ucian0) [reply qiskit/circuit/gate.py]: This is a question for @eddieschoute

File: qiskit/circuit/gate.py [hunk 9]
```diff
@@ -42,6 +47,28 @@ def to_matrix(self):
         """
         raise QiskitError("to_matrix not defined for this {}".format(type(self)))
 
+    def power(self, exponent):
```
[the remark this answers is not in the document]
Comment (1ucian0) [reply qiskit/circuit/gate.py]: Added in 076edb62

File: qiskit/circuit/gate.py [hunk 10]
```diff
@@ -42,6 +47,28 @@ def to_matrix(self):
...
+        # Raise the diagonal entries to the specified power
+        decomposition_power = list()
+        for element in decomposition.diagonal():  # assert off-diagonal are 0
+            decomposition_power.append(pow(element, exponent))
+        # Then reconstruct the resulting gate.
+        unitary_power = unitary @ np.diag(decomposition_power) @ unitary.conj().T
+        sub_instruction = UnitaryGate(unitary_power)
+
+        instruction = Instruction(name="%s^%s" % (self.name, exponent), num_qubits=self.num_qubits,
+                                  num_clbits=self.num_clbits, params=self.params)
```
[the remark this answers is not in the document]
Comment (1ucian0) [reply qiskit/circuit/gate.py]: See https://github.com/Qiskit/qiskit-terra/pull/2811/files#r309495337

File: test/python/circuit/test_instruction_power.py (hunk 4 above)
Comment (1ucian0) [reply test/python/circuit/test_instruction_power.py]: Test added in d26a05ca.

It seems that `-0.2` is not working @eddieschoute 

```
AssertionError: 
Not equal to tolerance rtol=1e-07, atol=0

x and y nan location mismatch:
 x: array([[1.      +0.j      ,      nan     +nanj],
       [     nan     +nanj, 0.951057-0.309017j]])
 y: array([[1.      +0.j      , 0.      +0.j      ],
       [0.      +0.j      , 0.951057-0.309017j]])
```

File: test/python/circuit/test_instruction_power.py (hunk 5 above)
Comment (1ucian0) [reply test/python/circuit/test_instruction_power.py]: in the test description?

File: qiskit/circuit/gate.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (eddieschoute) [reply qiskit/circuit/gate.py]: If it's not, some bad numerical error must've happened, perhaps if the eigenvalues are close to zero for example. The resulting gate would not be valid so it definitely should be reported somehow.

File: test/python/circuit/test_instruction_power.py (hunk 5 above)
Comment (kdk) [reply test/python/circuit/test_instruction_power.py]: An inline comment would be enough.

File: test/python/circuit/test_instruction_power.py (hunk 4 above)
Comment (eddieschoute) [reply test/python/circuit/test_instruction_power.py]: There are NaNs in there so something was divided by zero probably

File: test/python/circuit/test_instruction_power.py (hunk 5 above)
Comment (eddieschoute) [reply test/python/circuit/test_instruction_power.py]: I'd just use matrix power since that is semantically what I meant to do.

File: qiskit/circuit/instruction.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (kdk) [reply qiskit/circuit/instruction.py]: Good point, `type(self)(...)` is not what we want.

It would be nice though to be keep the return as a `Gate`, since they're assumed to be unitary. Otherwise, code like `op.power(x).to_matrix()` would be invalid, even if `op` is a `Gate`.

File: qiskit/circuit/instruction.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (eddieschoute) [reply qiskit/circuit/instruction.py]: Why not use the identity gate?

File: test/python/circuit/test_instruction_power.py (hunk 4 above)
Comment (kdk) [reply test/python/circuit/test_instruction_power.py]: The `NaN`s are from `SGate().to_matrix() ** exponent`

File: qiskit/circuit/instruction.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (kdk) [reply qiskit/circuit/instruction.py]: `IdGate` is defined to be a wait gate. (ref. https://github.com/Qiskit/qiskit-terra/issues/2373#issuecomment-504089670 )

[the remark this answers is not in the document]
Comment (1ucian0) [reply qiskit/circuit/instruction.py]: because `iden` means "wait" and is not optimized.

File: test/python/circuit/test_instruction_power.py (hunk 4 above)
Comment (1ucian0) [reply test/python/circuit/test_instruction_power.py]: Fixed in e6069b4e

File: qiskit/circuit/gate.py [hunk 11]
```diff
@@ -42,6 +47,41 @@ def to_matrix(self):
...
+        Zero exponent creates an instruction with u3(0,0,0).
+
+        Args:
+            exponent (float): Gate^exponent
+
+        Returns:
+            Instruction: Containing the definition.
+        """
+        if int(exponent) == exponent:
+            return super().power(exponent)
```
Comment (eddieschoute) [inline qiskit/circuit/gate.py]: I think the exponent should be cast to an int here.

Comment (1ucian0) [reply qiskit/circuit/gate.py]: It does inside `Instruction.power`

File: qiskit/circuit/instruction.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (1ucian0) [reply qiskit/circuit/instruction.py]: Once we get composite gates as gates, we can do this https://github.com/Qiskit/qiskit-terra/issues/2896

File: test/python/circuit/test_instruction_power.py (hunk 5 above)
Comment (1ucian0) [reply test/python/circuit/test_instruction_power.py]: `matrix_power` only works with ints, righ?

**Comment by itoko** [comment]
Calling integer exponentiation `Instruction.power` sounds strange for me. Why don't we call it `Instruction.repeat` (It can have the same implementation as Gate, different from `power`)?

File: test/python/circuit/test_instruction_power.py (hunk 5 above)
Comment (kdk) [reply test/python/circuit/test_instruction_power.py]: `scipy.linalg.fractional_matrix_power`?

File: qiskit/circuit/gate.py (hunk 11 above)
Comment (eddieschoute) [reply qiskit/circuit/gate.py]: Ok, but technically Instruction.power only takes integers as inputs so this is wrong type-wise.

File: test/python/circuit/test_instruction_power.py (hunk 5 above)
Comment (kdk) [reply test/python/circuit/test_instruction_power.py]: Though maybe it would be better to (additionally?) test some of the general exponential properties we expect to hold (e.g. `(op^(n))^(1/n) == op`, `op^(n) * op^(-n) = I`, ...), since we use something like `fractional_matrix_power` under the hood anyway.

File: qiskit/circuit/instruction.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (1ucian0) [reply qiskit/circuit/instruction.py]: Solved in 925cc51d

**Comment by 1ucian0** [comment]
> Calling integer exponentiation `Instruction.power` sounds strange for me. Why don't we call it `Instruction.repeat` (It can have the same implementation as Gate, different from `power`)?

What it means `Instruction.repeat(-2)`? If you need repeat, it can be added. But it's not exactly the same...

File: test/python/circuit/test_instruction_power.py (hunk 5 above)
Comment (1ucian0) [reply test/python/circuit/test_instruction_power.py]: Comment added in 3ab220a5

Comment (1ucian0) [reply test/python/circuit/test_instruction_power.py]: invariant test (`(op^(1/n))^(n) == op`) added in 653e4ab1

**Comment by itoko** [comment]
> > Calling integer exponentiation `Instruction.power` sounds strange for me. Why don't we call it `Instruction.repeat` (It can have the same implementation as Gate, different from `power`)?
> 
> What it means `Instruction.repeat(-2)`? If you need repeat, it can be added. But it's not exactly the same...

No. I means `Instruction.repeat` defined for *positive* integer would be enough and better. What `Instruction.power(-2)` means? I need `Gate.power(-2)`, but I don't understand the meaning and necessity of `Instruction.power(-2)`.

**Comment by 1ucian0** [comment]
Right now, custom gates are instruction (to be changed, #2896). `Instruction.power(-2)` of `-S-S-` should be `-Sdg-Sdg-Sdg-Sdg-`. Or am I missing something?

**Comment by kdk** [comment]
I think it’s okay for now for `CompositeInstruction`s to only support repeat. When #2896 makes `CompositeGate`s (including implementing their `to_matrix`), they can implement a `power` method. (This will also probably require moving `Barrier` from an `Instruction` to a `Directive`.)

But for now, `Instruction`s having `repeat` and `Gate`s having `power` is consistent with both how things are laid out now, and the direction we’d like to head.

**Comment by 1ucian0** [comment]
from @ajavadia 
> yeah i think repeat and power are both valuable. repeat just saves you from writing a loop. power is smart. We should allow gates to override their `.power` method. Like an `Adder(5).power(2)` should be `Adder(10)` not `Adder(5).Adder(5)`

File: qiskit/circuit/gate.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (ajavadia) [reply qiskit/circuit/gate.py]: looks like this check should be done @1ucian0

File: test/python/circuit/test_gate_power.py [hunk 12]
```diff
@@ -0,0 +1,257 @@
...
+                          [-0.70710678118, 0.70710678118]], dtype=complex)
+
+        result = UnitaryGate([[0, 1j], [-1j, 0]]).power(1 / 2)
+
+        self.assertEqual(result.label, 'unitary^0.5')
+        self.assertEqual(len(result.definition), 1)
+        self.assertIsInstance(result, Gate)
+        assert_allclose(result.definition[0][0].to_matrix(), expected)
+
+    def test_starndard_sqrt(self):
```
Comment (ajavadia) [inline test/python/circuit/test_gate_power.py]:
```suggestion
    def test_standard_sqrt(self):
```

File: test/python/circuit/test_gate_power.py [hunk 13]
```diff
@@ -0,0 +1,257 @@
...
+        qr = QuantumRegister(1, 'qr')
+        circuit = QuantumCircuit(qr)
+        circuit.append(SGate().power(1 / n).power(n), [qr[0]])
+        result = PassManager([Unroller('u3'), Optimize1qGates()]).run(circuit)
+
+        expected_circuit = QuantumCircuit(qr)
+        expected_circuit.append(SGate(), [qr[0]])
+        expected = PassManager([Unroller('u3'), Optimize1qGates()]).run(expected_circuit)
+
+        self.assertEqual(result, expected)
```
Comment (ajavadia) [inline test/python/circuit/test_gate_power.py]: why compare circuits here and go through transpile? can just check the `gate.to_matrix()` like every other test here.

File: qiskit/circuit/gate.py [hunk 14]
```diff
@@ -42,6 +45,26 @@ def to_matrix(self):
...
+    def power(self, exponent):
+        """Creates a unitary gate as `gate^exponent`.
+
+        Args:
+            exponent (float): Gate^exponent
+
+        Returns:
+            UnitaryGate: To which `to_matrix` is self.to_matrix^exponent.
+        """
+        from qiskit.extensions.unitary import UnitaryGate  # pylint: disable=cyclic-import
```
Comment (ajavadia) [inline qiskit/circuit/gate.py]: I'm a bit concerned by this class hierarchy. The `Gate` class is now aware of its derived class `UnitaryGate`. So if one day you decide to remove UnitaryGate, you break its parent. This is against object oriented principles. But I don't have a better solution now.

File: test/python/circuit/test_gate_power.py [hunk 15]
```diff
@@ -0,0 +1,257 @@
...
+    def test_sgate_float_minus_0_1(self, n=-0.1):
+        """Test Sgate.power(-0.1) method.
+        """
+        result = SGate().power(n)
+
+        expected = array([[1, 0],
+                          [0, 0.98768834 - 0.15643447j]], dtype=complex)
+        self.assertEqual(result.label, 's^%s' % n)
+        self.assertIsInstance(result, UnitaryGate)
+        assert_array_almost_equal(result.to_matrix(), expected)
```
Comment (ajavadia) [inline test/python/circuit/test_gate_power.py]: There is a lot of repetitions in these tests. Can you use `ddt` for all of them? 250 lines of test is a bit extreme for testing a simple feature like this and it will make it hard to parse through these tests later. I understand that the ground truth matrix has been hardcoded in these and that's why each one is a different test case, but I think you can just use `matrix_power` to compute them like you have done for the other `ddt` cases. Presumably that's what you have done anyway to get the ground truths.

File: qiskit/circuit/gate.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (nonhermitian) [reply qiskit/circuit/gate.py]: I will point out that all of this boils down to whether the input is a unitary or not. The non-diagonals being zero falls directly from this. The eigenvales or a unitary are necessarily nonzero, |lambda|=1 so that should be fine. It might make more sense to verify a gate is unitary up to some tolerance to begin with.

File: qiskit/circuit/gate.py (hunk 14 above)
Comment (1ucian0) [reply qiskit/circuit/gate.py]: Should standard gates subclass `UnitaryGate`? The method can be in `UnitaryGate` in that case.

File: test/python/circuit/test_gate_power.py (hunk 15 above)
Comment (1ucian0) [reply test/python/circuit/test_gate_power.py]: I moved to ddt in acb5232a. `matrix_power` only works with integers. Have a look.

File: test/python/circuit/test_gate_power.py (hunk 13 above)
Comment (1ucian0) [reply test/python/circuit/test_gate_power.py]: probably is a left over from a different iteration on the feature. Fixed in 7cf5d61f

File: test/python/circuit/test_gate_power.py (hunk 15 above)
Comment (1ucian0) [reply test/python/circuit/test_gate_power.py]: Similarly, d0807be5.

Comment (1ucian0) [reply test/python/circuit/test_gate_power.py]: currently 183 lines...

File: qiskit/circuit/gate.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (1ucian0) [reply qiskit/circuit/gate.py]: I added the check (please, have a look if it makes sense or if it's inefficient) in cb1f4e8e

I could not trigger the check tho. Any example?

File: qiskit/circuit/gate.py (hunk 14 above)
Comment (ajavadia) [reply qiskit/circuit/gate.py]: let's leave it at this for now, but clearly we need better class hierarchy.

**Comment by ajavadia** [comment]
Looks good, it just has to pass CI.

**Comment by 1ucian0** [comment]
@ajavadia says: 
> Looks good, it just has to pass CI.

Done in https://github.com/Qiskit/qiskit-terra/pull/2811/commits/176ac56aab046dc4803498dce5b08861d20b4649
